Dockerfile

```
@@ -1,41 +1,63 @@
-# NOTE: Testing Only. DO NOT use this in production
-
-ARG NODE_VERSION=16-alpine
-
-FROM node:${NODE_VERSION}
-
-# Required to run OracleDB
-# Technically not required for the others, but I'd rather have 1 image that works for all, instead of building n images
-# per test
-#WORKDIR /tmp
-#RUN apk --no-cache add libaio libnsl libc6-compat curl && \
-#    curl -o instantclient-basiclite.zip https://download.oracle.com/otn_software/linux/instantclient/instantclient-basiclite-linuxx64.zip -SL && \
-#    unzip instantclient-basiclite.zip && \
-#    mv instantclient*/ /usr/lib/instantclient && \
-#    rm instantclient-basiclite.zip && \
-#    ln -s /usr/lib/instantclient/libclntsh.so.19.1 /usr/lib/libclntsh.so && \
-#    ln -s /usr/lib/instantclient/libocci.so.19.1 /usr/lib/libocci.so && \
-#    ln -s /usr/lib/instantclient/libociicus.so /usr/lib/libociicus.so && \
-#    ln -s /usr/lib/instantclient/libnnz19.so /usr/lib/libnnz19.so && \
-#    ln -s /usr/lib/libnsl.so.2 /usr/lib/libnsl.so.1 && \
-#    ln -s /lib/libc.so.6 /usr/lib/libresolv.so.2 && \
-#    ln -s /lib64/ld-linux-x86-64.so.2 /usr/lib/ld-linux-x86-64.so.2
-#
-#ENV ORACLE_BASE /usr/lib/instantclient
-#ENV LD_LIBRARY_PATH /usr/lib/instantclient
-#ENV TNS_ADMIN /usr/lib/instantclient
-#ENV ORACLE_HOME /usr/lib/instantclient
+# syntax=docker/dockerfile:1.4
 
-WORKDIR /directus
+####################################################################################################
+## Prepare Workspace
+
+FROM node:18-alpine AS workspace
+WORKDIR /workspace
+
+RUN corepack enable && corepack prepare pnpm@7.30.0 --activate
+
+COPY pnpm-lock.yaml .
+RUN --mount=type=cache,id=pnpm,target=/root/.local/share/pnpm/store pnpm fetch
 
 COPY . .
+RUN pnpm install --recursive --frozen-lockfile
+
+####################################################################################################
+## Create Deployment
 
-RUN apk add --update python3 make g++\
-   && rm -rf /var/cache/apk/*
+FROM workspace AS pruned
+WORKDIR /workspace
+ENV NODE_OPTIONS=--max-old-space-size=8192
 
-RUN npm install
+RUN pnpm --recursive run build
+RUN pnpm --filter directus deploy --prod pruned
 
-WORKDIR /directus/api
+####################################################################################################
+## Create Production Image
 
-CMD ["sh", "-c", "node ./cli.js bootstrap && node ./dist/start.js;"]
-EXPOSE 8055/tcp
+FROM node:18-alpine
+WORKDIR /directus
+EXPOSE 8055
+
+ENV \
+    DB_CLIENT="sqlite3" \
+    DB_FILENAME="/directus/data/database/database.sqlite" \
+    EXTENSIONS_PATH="/directus/data/extensions" \
+    STORAGE_LOCAL_ROOT="/directus/data/uploads" \
+	 NODE_ENV=production
```

I think the counterpart for `ENV` to
```
RUN : \
  && ... \
  ;
```

would be this format:

```suggestion
ENV DB_CLIENT="sqlite3"
ENV DB_FILENAME="/directus/data/database/database.sqlite"
ENV EXTENSIONS_PATH="/directus/data/extensions"
ENV STORAGE_LOCAL_ROOT="/directus/data/uploads"
ENV NODE_ENV="production"
```

---

If I'm not mistaken, the reason to do it the "ugly" way is because each command in a Dockerfile creates a new cache "layer", so doing 5 `ENV` calls is way less efficient than doing 1 with this multi-line approach 👍🏻

---

While this was indeed the case in the past, it shouldn't be an issue anymore:

> Only the instructions `RUN`, `COPY`, `ADD` create layers. Other instructions create temporary intermediate images, and don’t increase the size of the build.

(From https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#minimize-the-number-of-layers)